mkosi/distributions/gentoo.py

```diff
@@ -16,45 +17,103 @@
+
+thread_counts = (os.cpu_count() or 1) * 2
 
 
 def invoke_emerge(
     state: MkosiState,
+    sysroot: Optional[Path] = None,
+    root: Optional[Path] = None,
+    bwrap_params: list[PathString] = [],
     pkgs: Sequence[str] = (),
     actions: Sequence[str] = (),
     opts: Sequence[str] = (),
+    env: dict[str, str] = {},
 ) -> None:
-    jobs = os.cpu_count() or 1
+    bwrap: list[PathString] = []
+    if sysroot is not None:
+        target_root_mntp = "/tmp/mkosi-root"
+        bwrap += ["--bind", state.root, target_root_mntp]
+        root = Path(target_root_mntp)
+    else:
+        sysroot = state.root
+        root = None
+
     emerge_default_opts = [
         "--buildpkg=y",
         "--usepkg=y",
         "--keep-going=y",
-        f"--jobs={jobs}",
-        f"--load-average={jobs+1}",
+        f"--jobs={thread_counts}",
+        f"--load-average={thread_counts+1}",
         "--nospinner",
+        *([f"--root={root}"] if root else []),
     ]
     if ARG_DEBUG.get():
         emerge_default_opts += ["--verbose", "--quiet=n", "--quiet-fail=n"]
     else:
         emerge_default_opts += ["--quiet-build", "--quiet"]
     cmd = ["emerge", *pkgs, *emerge_default_opts, *opts, *actions]
-    run_workspace_command(state.root, cmd, network=True, env=state.environment)
+    bwrap += [
+        "--bind", state.cache_dir / "binpkgs", "/var/cache/binpkgs",
+        "--bind", state.cache_dir / "distfiles", "/var/cache/distfiles",
+        "--bind", state.cache_dir / "repos", "/var/db/repos",
+        *bwrap_params
+    ]
+    run_workspace_command(sysroot, cmd, bwrap_params=bwrap, network=True,
+                          env=env)
+
+
+@dataclass
+class GentooConstants:
+    gentoo_mirrors: str
+    portdir: Path
+    binrepos_conf_file: Path
+    ebuild_sh_env_dir: Path
+    user_config_path: Path
+
+
+@dataclass
+class GentooAtoms:
+    stage2: List[str]
+    bare_minimal: List[str]
+    boot: List[str]
+    system: List[str]
 
 
 class Gentoo:
     arch_profile: Path
     arch: str
-    custom_profile_path: Path
-    ebuild_sh_env_dir: Path
     emerge_vars: dict[str, str]
-    portage_cfg_dir: Path
-    profile_path: Path
+    user_config_path: Path
     root: Path
-    pkgs: dict[str, list[str]] = {}
+    stage3_cache: Path
+    stage3_tar_path: Path
+    pkgs: GentooAtoms = GentooAtoms(
+        stage2=[
+            "sys-apps/baselayout",
+            "sys-apps/util-linux",
+        ],
+        bare_minimal=[
+            "app-shells/bash",
+            "sys-apps/systemd",
+        ],
+        boot=[
+            "sys-kernel/gentoo-kernel-bin",
+        ],
+        system=["@system"]
+    )
+    portage_consts: GentooConstants = GentooConstants(
+        gentoo_mirrors="http://distfiles.gentoo.org",
+        portdir=Path("/var/db/repos/gentoo"),
+        binrepos_conf_file=Path("etc/portage/binrepos.conf"),
+        ebuild_sh_env_dir=Path("etc/portage/env"),
+        user_config_path=Path("etc/portage")
```

Comment:
I'm sorry if that was unclear: What is the benefit of defining a dataclass and encapsulating these constants in it, if those are just four values that will likely never change? I don't think it make it improves readability to have `portage_consts.user_config_path`, when it could be `USER_CONFIG_PATH` for some `USER_CONFIG_PATH=Path("etc/portage")` at the top of the file.